### ISSUE TITLE
Jobs hung in READY

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#601]
In GitLab by @mcastril on Oct 15, 2020, 13:44

Hi @dbeltrankyl . @ mfalls has this workflow with a wrapper in a34f 

http://bscesweb04.bsc.es/autosubmitapp/experiment/a34f

It seems there was a problem when some of the inner jobs failed.

```
2020-10-13 19:04:59,390 Command sacct -n -j 11971941 -o JobId%25,State,NCPUS,NNodes,Submit,Start,End,ConsumedEnergy,MaxRSS%25,AveRSS%25 in mn-bsc32 successful with out message:                  11971941     FAILED          2        1 2020-10-13T19:01:37 2020-10-13T19:03:09 2020-10-13T19:03:21          1.76K
           11971941.batch     FAILED          2        1 2020-10-13T19:03:09 2020-10-13T19:03:09 2020-10-13T19:03:21          1.35K                      815K                      815K
          11971941.extern  COMPLETED          2        1 2020-10-13T19:03:09 2020-10-13T19:03:09 2020-10-13T19:03:22          1.76K                      538K                      538K

2020-10-13 19:04:59,390 Output                  11971941     FAILED          2        1 2020-10-13T19:01:37 2020-10-13T19:03:09 2020-10-13T19:03:21          1.76K                                                  
           11971941.batch     FAILED          2        1 2020-10-13T19:03:09 2020-10-13T19:03:09 2020-10-13T19:03:21          1.35K                      815K                      815K
          11971941.extern  COMPLETED          2        1 2020-10-13T19:03:09 2020-10-13T19:03:09 2020-10-13T19:03:22          1.76K                      538K                      538K

2020-10-13 19:04:59,413 Updating FAILED jobs
2020-10-13 19:04:59,413 Resetting job: a34f_19600101_fc0000_1_SIM status to: READY for retrial...
2020-10-13 19:04:59,413 Resetting job: a34f_19600101_fc0000_2_SIM status to: READY for retrial...
2020-10-13 19:04:59,413 Update finished
2020-10-13 19:04:59,413 Updating WAITING jobs
2020-10-13 19:04:59,413 Update finished
2020-10-13 19:04:59,413
Job Name                           Job Id         Job Status     Job Platform        Job Queue
2020-10-13 19:04:59,414 a34f_INI                           11971927       COMPLETED      marenostrum4        debug
2020-10-13 19:04:59,417 Saving JobList: /esarchive/autosubmit/a34f/pkl/job_list_a34f.pkl
2020-10-13 19:04:59,418 Job list saved
2020-10-13 19:04:59,420
Jobs ready for marenostrum4: 1
2020-10-13 19:04:59,420 Number of jobs available: 6
2020-10-13 19:04:59,421 Number of jobs ready: 1
2020-10-13 19:04:59,421 Jobs ready for marenostrum4: 1
2020-10-13 19:04:59,431 Custom directives from platform.conf: None
2020-10-13 19:04:59,655 a34f_19600101_fc0000_2_SIM_STAT have been removed
2020-10-13 19:05:00,805 Command /gpfs/scratch/bsc32/bsc32786/a34f/LOG_a34f/submit_marenostrum4.sh in mn-bsc32 successful with out message: Submitted batch job 11972003

2020-10-13 19:05:00,805 Output Submitted batch job 11972003
```

As you can see, jobs were moved to READY, but after that it seems it is not finding any ready job:

```
6 of 7 jobs remaining (19:05)
2020-10-13 19:05:20,880 Sleep: 10
2020-10-13 19:05:20,880 Number of retrials: 3
2020-10-13 19:05:20,881 WRAPPER CHECK TIME = 10
2020-10-13 19:05:20,881 Updating FAILED jobs
2020-10-13 19:05:20,881 Update finished
2020-10-13 19:05:20,881 Updating WAITING jobs
2020-10-13 19:05:20,882 Update finished
2020-10-13 19:05:20,882
Jobs ready for marenostrum4: 0
2020-10-13 19:05:20,882 Number of jobs available: 6
2020-10-13 19:05:20,882 Number of jobs ready: 0
2020-10-13 19:05:20,885 Updating FAILED jobs
2020-10-13 19:05:20,885 Update finished
2020-10-13 19:05:20,885 Updating WAITING jobs
2020-10-13 19:05:20,885 Update finished
2020-10-13 19:05:30,898 Reloading parameters...
2020-10-13 19:05:30,912 Loading parameters...
2020-10-13 19:05:30,913 Loading project parameters...
2020-10-13 19:05:30,914

```

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 15, 2020, 13:46

@ mfalls , you can maybe do a backup of `/esarchive/autosubmit/a34f/pkl` so @dbeltrankyl can take a look to that, and then start from scratch because it seems Autosubmit failed when one of those SIM failed.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 15, 2020, 13:46

Hi, 

Is this on 3.13.0b? I believe that is related to #599 and a fix is incoming

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 15, 2020, 13:51

yes it is

**Comment by kinow** [comment]
In GitLab by @ lpalma on Oct 16, 2020, 13:22

Hi @mcastril @dbeltrankyl, 

I'm not sure if it is related to this, but our operational experiment (a2on) got stuck on the first job (DL) tonight. I have checked the logs, and this is the error prompted:


```
23 of 23 jobs remaining (23:02)
Job a2on_20201015_DL is RUNNING


23 of 23 jobs remaining (23:02)
^[[31m[ERROR] Can not send command nohup kill -0 9831 >& /dev/null; echo $? to bscearth000: SSH session not active^[[0m^[[39m
^[[33m[WARNING] Retrying check job command: nohup kill -0 9831 >& /dev/null; echo $?^[[0m^[[39m
^[[31m[ERROR] Can not get job status for job id (9831), retrying in 10 sec^[[0m^[[39m
```

The job was being run on the bsc000 machine.

Thanks in advance,

@pabretonniere

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 16, 2020, 13:51

Hi, 
Yes seems that bsc000 and ecaccess platforms has some issues lately, as discussed on #600. 

The issue seems that AS 3.12.0 hangs if some of these platforms have connection issues. Since 3.12.0 is long stable we won't touch this since 3.13.0b will cover these kinds of bugs and will be in a short future the stable one. 

If you have this issue on 3.12.0, you have to stop autosubmit and run it again, if the job is not on sacct history anymore, you will have to do an autosubmit recovery to check if that job is really completed, alternatively you can use 3.13.0b although it is still a beta version so you may encounter bugs.

**Comment by kinow** [comment]
In GitLab by @ lpalma on Oct 16, 2020, 14:17

Ok, thanks for the clarification @dbeltrankyl!!

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 16, 2020, 16:55

Thanks @dbeltrankyl . Only clarify that this issue (#600) is about a manual migration + recovery. You maybe refer to #568 or https://earth.bsc.es/gitlab/es/auto-ecearth3/-/issues/1427

**Comment by kinow** [comment]
In GitLab by @etiennesky on Oct 18, 2020, 07:36

Hi @dbeltrankyl can you please consider fixing this quite serious issue in 3.12.0 ? 

You say bugfixes should go in 3.13.0 but since you cannot guarantee is doesn't have any bugs, that is not an option for running production experiments as the ones in https://earth.bsc.es/gitlab/es/auto-ecearth3/-/issues/1427

This bug is a showstopper for me since I have to kill AS more than twice a day to get my experiments running.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 18, 2020, 17:40

@etiennesky the connection issue is mostly affecting CCA experiments, and this is because there is an ongoing issue in CCA connectivity (we have been testing 3.12.0 in CCA without issues till now and it passed the "test period" that took several months).

I think it would be less dangerous that you guys use 3.13.0 while ecaccess is being fixed, that changing 3.12.0 that is used also for MN4 workflows and hasn't been touched in months. That's for example what the AC guys have been doing with Data assimilation experiments (because they have specific workflow necessities, not relevant for the rest).

This changes are not trivial, more than a bugfix we could say that it's part of the major rework to make Autosubmit more stable.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 18, 2020, 18:07

In case there was an easy fix/mitigation we could either create an alternative 3.12.0-fix version, but I don't think they are... what do you think @dbeltrankyl ?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 19, 2020, 09:38

I can take a look without doing deep changes but instead of AS hang, maybe the jobs will be put in unknown status instead of hang AS I have to check, it will take some time tho due to the version being much different to the current 3.13.0

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 20, 2020, 16:37

Related to #601 ,#600 https://earth.bsc.es/gitlab/es/auto-ecearth3/-/issues/1427#note_100744

I created a patch on  a new module called `autosubmit/3.12.0-fixed-foss-2015a-Python-2.7.9`, 

Please, the users that had issues with 3.12.0 current version, swap to this version to see if the changes are correct and do not break anything if it goes well it will be the default one. 

The module is essentially the same than 3.12.0 but instead of hanging all the experiment what will now occur now in these cases that job, in particular, will be set into UNKNOWN status. if you have the mail notifier set on, it will send you an email whenever this issue happens. 

If autosubmit can detect later the real status of the job it will try to put the correct status, if not you will need to do it manually. 


Apologies for the issues caused, this is just a non-invasive patch, further improvements will be on 3.13.0b

**Comment by kinow** [comment]
In GitLab by @etiennesky on Oct 20, 2020, 19:52

Hi @dbeltrankyl @mcastril I am having this issue now with AS 3.12.0 (the default, I did not change any modules) with expid a36z

```
Job a36z_20150101_fc2_30_TRANSFER is COMPLETED
[CRITICAL] Unhandled exception on Autosubmit: 'in <string>' requires string as left operand, not int
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.12.0-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0-py2.7.egg/autosubmit/autosubmit.py", line 416, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/shared/earth/software/autosubmit/3.12.0-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0-py2.7.egg/autosubmit/autosubmit.py", line 1056, in run_experiment
    if Status.VALUE_TO_KEY[job.status] in job.notify_on or Status.UNKNOWN in Status.VALUE_TO_KEY[job.status]:
TypeError: 'in <string>' requires string as left operand, not int
```

looks like you updated 3.12.0 cause the date is from today

```
ls -l /shared/earth/software/autosubmit/3.12.0-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0-py2.7.egg/autosubmit/autosubmit.py
-rw-r--r-- 1 eadmin Earth 158248 Oct 20 16:20 /shared/earth/software/autosubmit/3.12.0-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0-py2.7.egg/autosubmit/autosubmit.py
```

I tried with the new module and I get the same error

```

Job a36z_20150101_fc2_30_TRANSFER is COMPLETED
[CRITICAL] Unhandled exception on Autosubmit: 'in <string>' requires string as left operand, not int
Traceback (most recent call last):
  File "/shared/earth/software/autosubmit/3.12.0-fixed-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0-py2.7.egg/autosubmit/autosubmit.py", line 416, in parse_args
    return Autosubmit.run_experiment(args.expid, args.notransitive)
  File "/shared/earth/software/autosubmit/3.12.0-fixed-foss-2015a-Python-2.7.9/lib/python2.7/site-packages/autosubmit-3.12.0-py2.7.egg/autosubmit/autosubmit.py", line 1056, in run_experiment
    if Status.VALUE_TO_KEY[job.status] in job.notify_on or Status.UNKNOWN in Status.VALUE_TO_KEY[job.status]:
TypeError: 'in <string>' requires string as left operand, not int


```

Note that a376 is currently running  (also on cca) without problems...

**Comment by kinow** [comment]
In GitLab by @etiennesky on Oct 20, 2020, 19:58

mentioned in issue auto-ecearth3#1427

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 20, 2020, 20:25

Sorry for this @etiennesky . It looks evident to me too that something was wrong in the update. @dbeltrankyl will confirm early tomorrow.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 20, 2020, 21:00

i'll check now

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 20, 2020, 21:13

hi @etiennesky ,

use 3.12.0-fixed there is an issue with a library(portalocker) on 3.12.0 working on it but 3.12.0-fixed should work now

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 20, 2020, 21:32

Hi @etiennesky ,

Fixed should work as expected,

default had to be updated to avoid the update of portalocker library, 
while the module update the default version went modified to 4.0b, I added back the link and should have looked the source of the reason about that error, apologies.

The late issues should be fixed now


edit: 

**The whole process I think it was:**

Update 3.12.0-fixed =>  failed due portarlocker, some part of the script pointed to both versions 3.12.0 and 3.12.0-fixed. then both got deleted without I notice ( guess it have *wild card?

Update 3.12.0-fixed => successful
Detected that default version doesn't point anymore to 4.0b, then I noticed that 3.12.0 went missing.
Updated 3.12.0 but I guess I didn't change the tag in both sides of the script, otherwise, It would crash due to the portalocker error.

About the issue:

The notifier part couldn't replicate the ecaccess error so I couldn't test that part and it just was a matter of syntax change in for == (dumb for myside, apologies)

the good news is that if that part crashed, the experiment isn't being hung anymore.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 21, 2020, 10:27

Thanks @dbeltrankyl.

> I am having this issue now with AS 3.12.0 (the default, I did not change any modules) with expid a36z

@etiennesky so in the end what it happened is that as you were guessing, 3.12.0-fixed and 3.12.0 were pointing to the new one. But the error that you reported was due to the loss of the connection, that in the patched version should be managed with the setting of the job to UNKNOWN. There was a bug in this new part that was fixed. The "positive" side, as @dbeltrankyl mentioned, is that instead of that crash, your Autosubmit would have entered in the "session not active" infinite loop, so it wouldn't have been any better.

This particular issue is very difficult to reproduce because it depends on the environment, and I don't know if this would help in this particular case, but I will open an issue that I had in the backlog to create a small testing suite for Autosubmit, with a small experiment from every Auto-model.

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 27, 2020, 12:01

closed
